python/src/worker/tasks/factory.py

```diff
@@ -36,27 +36,53 @@ class TaskFactory:
             GetDoubletScore,
             GetMitochondrialContent,
             MarkerHeatmap,
-            GetExpressionCellSets
+            GetExpressionCellSets,
         )
     }
 
     def __init__(self):
         self.count_matrix = CountMatrix()
         self.count_matrix.sync()
 
+    def _log_exception(task):
+        trace = (
+            f"Exception for task {task.__class__.__name__}:\n"
+            f"{traceback.format_exc()}"
+        )
+
+        error(trace)
+        xray_recorder.current_segment().add_exception(e, trace)
+
+        return trace
+
     def submit(self, msg):
         task = self._factory(msg)
 
         # Try to perform task. If fails, send back an error to the API.
         try:
             result = task.compute()
             return result
-        except Exception as e:
-            trace = f"Exception for task {task.__class__.__name__}:\n" \
-                    f"{traceback.format_exc()}"
 
-            error(trace)
-            xray_recorder.current_segment().add_exception(e, trace)
+        except RWorkerException as r:
+            trace = self._log_exception(task)
+
+            # Only send real traces in development.
+            if config.CLUSTER_ENV == "development":
+                result = Result(trace, error=True)
+            else:
+                result = [
+                    Result(
+                        {
+                            "error": e.message,
+                            "code": e.code,
+                        },
+                        error=True,
+                    )
+                ]
+            return result
+
+        except Exception as e:
```

Comment:
This is the same piece of code, but for 2 different errors, you can do something like this:

```
        except Exception as e:
            trace = self._log_exception(task)

            # Only send real traces in development.
            if config.CLUSTER_ENV == "development":
                return result = Result(trace, error=True)

            if isinstance(e, RWorkerException):
                return [
                    Result(
                        {
                            "error": e.message,
                            "code": e.code,
                        },
                        error=True,
                    )
                ]

            return [
                    Result(
                        "An unexpected error occurred while performing the work.",
                        error=True,
                    )
                ]
```

Reply:
Implemented